### PR TITLE
feat: optimally build partial trees for use with levelcachestore

### PR DESCRIPTION
**Pull request by cryptonemo**
feat: add env logging

fix: format and clippy

### CODE REVIEW

**Comment by cryptonemo** [comment]
@dignifiedquire The good news is that this resolves #49 

The bad news is that it doesn't yet address the new issue #52.  I think that should be a separate PR after this one (no version bump required here on merge).  Sound reasonable?  I think I still need some details on that anyway, as it's not clear where the base layer data will be coming from or how best to structure.  Might be good to sync/pair for that step.

**Comment by dignifiedquire** [comment]
> @dignifiedquire The good news is that this resolves #49

:tada:

> The bad news is that it doesn't yet address the new issue #52.

@nemo yes, #52 should be a follow up pr

File: Cargo.toml [hunk 1]
```diff
@@ -31,6 +31,8 @@ tempdir = "0.3.7"
 tempfile = "3.0.7"
 positioned-io = "0.2"
 failure = "0.1.2"
+log = "0.4.7"
+env_logger = "0.7.1"
```
Comment (dignifiedquire) [inline Cargo.toml]: the env_logger should be a dev dependency

File: src/merkle.rs [hunk 2]
```diff
@@ -452,70 +429,75 @@ impl<T: Element, A: Algorithm<T>, K: Store<T>> MerkleTree<T, A, K> {
...
+
+        // Calculate the data layer width (termed 'segment_width')
+        // that we need in order to build the partial tree required to
+        // build the proof, given the data configuration specified by
+        // 'levels'.
+        let segment_width = self.leafs / cached_leafs;
+        let segment_start = (i / segment_width) * segment_width;
+        let segment_end = segment_start + segment_width;
+
+        info!("leafs {}, total size {}, total height {}, cache_size {}, cached levels above base {}, \
```
Comment (dignifiedquire) [inline src/merkle.rs]: should be `debug!`

File: src/merkle.rs [hunk 3]
```diff
@@ -452,70 +429,75 @@ impl<T: Element, A: Algorithm<T>, K: Store<T>> MerkleTree<T, A, K> {
...
+
+        // Generate entire proof with access to the base data, the
+        // cached data, and the partial tree.
         let proof = self.gen_proof_with_partial_tree(i, levels, &partial_tree);
 
-        Ok(ProofAndTree {
-            proof,
-            merkle_tree: partial_tree,
-        })
+        info!(
```
Comment (dignifiedquire) [inline src/merkle.rs]: should be `debug` probably

**Pull request merged**